### ISSUE TITLE
Allow derives to `allow(non_ascii_idents)` at the non-crate level

### ISSUE THREAD

**Issue by joshlf** [rust-lang/rust#151025]
In https://github.com/google/zerocopy/issues/2880, a user reported that our use of non-ASCII identifiers in generated code triggers the `non_ascii_idents` lint in their codebase.

In https://github.com/google/zerocopy/pull/2882, we tried to solve this by emitting `allow(non_ascii_idents)` in derive-generated code, but this unfortunately has no effect, simply triggering this warning:

```text
warning: allow(non_ascii_idents) is ignored unless specified at crate level
  --> zerocopy-derive/tests/issue_2880_expanded.rs:77:9
   |
77 |         non_ascii_idents,
   |         ^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_attributes)]` (part of `#[warn(unused)]`) on by default
```

This "crate-level only" behavior is also true of a number of other lints ([for example](https://github.com/rust-lang/rust/blob/aefa10405d7b67b3780027484cb02c85d3a3bf36/compiler/rustc_lint/src/non_ascii_idents.rs)). This restriction prevents us from using non-ASCII identifiers in generated code, which we intentionally do in order to minimize the likelihood of identifier conflicts with surrounding code. It would be good to support this in derive-generated code.

We propose one of the following:
- Lift this restriction
- Lift this restriction *within code annotated with `#[automatically_generated]`*
- Implicitly suppress this lint within code annotated with `#[automatically_generated]`

cc @jswrenn

**Comment by nxsaken** [comment]
Nominating since it has been [brought up](https://hackmd.io/1fpf7RyOQeGrBlg9MHUYZA?edit=&stext=16469%3A96%3A1%3A1780513479%3A-ntdfW) at today's Rust for Linux meeting.

**Comment by joshlf** [comment]
https://github.com/rust-lang/rust/pull/157497 addresses this by going the minimal route: Just allowing `non_ascii_idents` elsewhere (not just at the crate top-level). It's just one option, so feel free to close if that's not the route we want to go.

**Comment by ojeda** [comment]
@rustbot label A-rust-for-linux

(It would allow us to remove a custom `zerocopy` patch in our vendored copy)

**Comment by joshtriplett** [comment]
In principle, it seems valid to want `non_ascii_idents` at a finer-grained scope. For instance, "most of this crate should be ASCII, but this one function that implements a complex math formula can use various math symbols and Greek".

But the actual problem this is trying to solve for zerocopy seems to be "get a fresh symbol that doesn't conflict with anything", and for *that*, please, *please* give us a lang proposal for a feature that does what you actually want, rather than a hacky workaround. :)

We'd *love* to ship a `gensym`/`fresh` mechanism. I'd be thrilled to read a proposal for such a mechanism.

**Comment by tmandry** [comment]
We talked about this in today's lang meeting. On the specific ask, we did not have a problem with making the lint work at the item level. I consider this a quality-of-implementation issue, not a lang issue, unless there was an explicit earlier decision not to honor the lint at the item or module level.
